lib/options.c

```diff
@@ -40,371 +40,249 @@
-        /* no options for this TCTI yet */
+    ADD_TCTI("abrmd", tpm2_tools_tcti_abrmd_init)
 #endif
-        {
-            .name    = "help",
-            .has_arg = no_argument,
-            .flag    = &common_opts->help,
-            .val     = true,
-        },
-        {
-            .name    = "verbose",
-            .has_arg = no_argument,
-            .flag    = NULL,
-            .val     = 'V',
-        },
-        {
-            .name    = "quiet",
-            .has_arg = no_argument,
-            .flag    = NULL,
-            .val     = 'Q',
-        },
-        {
-            .name    = "version",
-            .has_arg = no_argument,
-            .flag    = NULL,
-            .val     = 'v',
-        },
-        {
-            .name = NULL,
-            .has_arg = no_argument,
-            .flag = NULL,
-            .val = '\0'
-        },
+};
+
+bool tpm2_handle_options (int argc, char **argv, char **envp, tpm2_options  *tool_opts, tpm2_option_flags *flags, TSS2_TCTI_CONTEXT **tcti) {
+
+    bool result = false;
+
+    // TODO handle execute man
+    UNUSED(envp);
+
+    struct option long_options [] = {
+        { "tcti",    required_argument, NULL,   'T' },
+        { "help",     no_argument,       NULL,  'h' },
+        { "verbose",   no_argument,       NULL, 'v' },
+        { "quiet",     no_argument,       NULL, 'Q' },
+        { "version",   no_argument,       NULL, 'V' },
     };
-    /*
-     * Start by populating the provided common_opts_t structure with data
-     * provided in the environment. Whatever we get here will be overriden
-     * by stuff from argv.
-     */
-    get_common_opts_from_env (common_opts);
-    /*
-     * Keep getopt_long quiet when we see options that aren't in the 'common'
-     * category. Reset option processing.
-     */
-    char **new_argv = { NULL, };
-    int new_argc = 1;
-    extern int opterr, optind;
-
-    opterr = 0;
-    optind = 1;
-    new_argv = calloc (1, sizeof (char*));
-    if (new_argv == NULL) {
-        fprintf (stderr, "Failed to allocate memory for tool argument "
-                 "vector: %s\n", strerror (errno));
-        return 2;
+
+    char *tcti_opts = NULL;
+    char *tcti_name = TCTI_DEFAULT_STR;
+    char *env_str = getenv (TPM2TOOLS_ENV_TCTI_NAME);
+    tcti_name = env_str ? env_str : tcti_name;
+
+    /* handle any options */
+    tpm2_options *opts = tpm2_options_new("T:hvVQ",
+            ARRAY_LEN(long_options), long_options, NULL, NULL);
+    if (!opts) {
+        return false;
+    }
+
+    /* Get the options from the tool */
+    if (tool_opts) {
+        tpm2_options *tmp = tpm2_options_cat(opts, tool_opts);
+        if (!tmp) {
+            tpm2_options_free(opts);
+            tpm2_options_free(tool_opts);
+            return false;
+        }
+        opts = tmp;
     }
-    new_argv[0] = argv[0];
-    while ((c = getopt_long (argc, argv, arg_str, long_options, &option_index))
+
+    /* Parse the options, calling the tool callback if unknown */
```

Comment:
Hmmm, right now the tools expect that the common options are processed before the specific of the tools. For example the -Q and -V options since many tools use the LOG_* and TOOL_OUTPUT macros in their options handling code.

Reply:
That's true... not sure how to deal with that. I don't think LOG_INFO() is used in option handling. We can just say, if uses LOG_INFO during option processing, they may not be output.